adding mainnet deployment guide and other tweaks (#430)
helium/docs: docs/mine-hnt/validators/mainnet/deployment-guide.mdx

---
id: validators-mainnet-deployment-guide
title: Validator Deployment Guide - Mainnet
sidebar_label: Deployment Guide - Mainnet
slug: /mine-hnt/validators/mainnet/deployment-guide
---

import useBaseUrl from "@docusaurus/useBaseUrl";

This guide will cover the end-to-end process of deploying a Helium Validator on Mainnet. It assumes you're starting from scratch. At the end, you'll have a Helium Validator deployed on the Network.

:::warning

> Although staking HNT has officially been enabled on Mainnet, Validators are not yet managing the Consensus Group - and therefore are not yet earning HNT. This will be enabled approximately 10-14 days after Mainnet reaches 100 staked Validators.

Reviewer: ```suggestion
Although initial staking has officially been enabled on Mainnet, Validators are not yet elected into the Consensus Group - and therefore are not yet earning HNT. This will be enabled at a later date, after the core developers believe it's safe to activate the chain variables. The current expectation is that this will be possible about 10-14 days after Mainnet reaches 100 staked Validators.
```

:::

:::info

- If at any point during this process you need assistance or have general questions, please join the `#validator` channel on [Helium Discord](https://discord.gg/helium).

:::

## Prerequisites

To successfully deploy a Helium Validator, you'll need to do the following:

- Install the [Helium CLI Wallet](https://github.com/helium/helium-wallet-rs). You are welcome to build from source, but easiest to download the latest binary release.
- Deploy a virtual server on a cloud service like AWS, GCP, Azure, or Digital Ocean. Or deploy a physical server on which to run the Validator. [The recommended hardware and networking requirements can be found here](/mine-hnt/validators/requirements/).

## Create Wallet

> Once you have your [Helium CLI Wallet](https://github.com/helium/helium-wallet-rs) installed locally, it's time to create your Helium Wallet. Run the following command to create it. (This command format assumes you're using the executable. If you've built the wallet from source it'll look slightly different.)

Reviewer 2: ```suggestion
Once you have your [Helium CLI Wallet](https://github.com/helium/helium-wallet-rs) installed locally, it's time to create your Helium Wallet. Run the following command to create it. (This command format assumes you're using the executable. If you've built the wallet from source it'll look slightly different.) 

At any time use -h or --help to get more help for a command.
```

+--------------------+-----------------------------------------------------+
| Network            | mainnet                                             |
+--------------------+-----------------------------------------------------+
| Type               | ed25519                                             |
+--------------------+-----------------------------------------------------+
| Sharded            | false                                               |
+--------------------+-----------------------------------------------------+
| PwHash             | Argon2id13                                          |
+--------------------+-----------------------------------------------------+
| Balance            | 0.00000000                                          |
+--------------------+-----------------------------------------------------+
| DC Balance         | 0                                                   |
+--------------------+-----------------------------------------------------+
| Securities Balance | 0.00000000                                          |
+--------------------+-----------------------------------------------------+
```

## Each Validator Requires 10000 HNT

> Running a Validator on Mainnet requires a stake of `10000 HNT` per Validator. After you've created your wallet, transfer `10000 HNT` into it for each Validator you intend to run. (You should actually transfer slightly more than `10000 HNT` for each Validator to cover the transaction costs required for the staking and unstaking operations. An additional `2` HNT per Validator will be more than enough to cover this.) 

Reviewer: ```suggestion
Running a Validator on Mainnet requires a stake of `10000 HNT` per Validator. After you've created your wallet, transfer a little more than `10000 HNT` into it for each Validator you intend to run. (The extra will cover the transaction costs required for the staking and unstaking operations. An additional `2` HNT per Validator will be more than enough to cover this.) 
```

Reviewer 2: Can we make the content that describes transaction costs a "Note:" to highlight it.

Author: Yeah. I'll update it.

```
grep avx /proc/cpuinfo
```

If nothing is returned from this command, your host system does not have AVX support and your Validator may not be stable.

:::

You have two options for deploying the Validator. You just need to do ONE of the following:

- [Deploy the Validator Using Docker](#deploy-the-miner-using-docker). Recommended if you are unfamiliar with building software.
- [Build the Validator form Source](#deploy-miner-from-source)

### Deploy the Validator Using Docker - RECOMMENDED

Start by updating your packager manager registry:

```
sudo apt-get update
```

And install Docker itself. (If needed, full directions on installing Docker on Ubuntu [can be found here](https://docs.docker.com/engine/install/ubuntu/).)

```
sudo apt-get install docker.io
```

To avoid needing to use Docker with `sudo` privileges, add your user to the docker group, replacing $USER with your username here:

```
sudo usermod -aG docker $USER
```

Log in and out of your account to apply these changes. You are now ready to use Docker.

Now that docker is installed and ready, let's run the Validator.

#### Run the Docker Container

> Before running the container for the first time, it is mandatory to pick a "system mount point". This allows you to easily maintain your Validator's blockchain identity - specifically it `swarm keys` - and blockchain state through image updates. Without doing this, you will lose these items when upgrading docker images.

Reviewer 2: ```suggestion
Before running the container for the first time, it is mandatory to pick a "system mount point". This allows you to easily maintain your Validator's blockchain identity - specifically `swarm keys` - and blockchain state through image updates. Without doing this, you will lose these items when upgrading docker images.
```

The `validator` directory should be created on a persistent EBS volume (if you're using AWS) or something similar that lives across server lifetimes.

```
mkdir $HOME/validator_data
```

You can then use the `run` command to start your container for the first time:

```
docker run -d --init \
--restart always \
--publish 2154:2154/tcp \
--name validator \
--mount type=bind,source=$HOME/validator_data,target=/var/data \

> quay.io/team-helium/validator:latest-val-amd64

Reviewer: Need to verify this tag

> quay.io/team-helium/validator:latest-val-amd64

Reviewer 2: ```suggestion
quay.io/team-helium/validator:latest-validator-amd64
```

- `--restart always` asks Docker to keep the image running, starting the image on boot and restarting the image if it crashes. Depending on how you installed Docker in your system, it'll start on boot. In the AWS AMI above, we use systemd (systemctl status docker to check).
- `--publish 2154:2154/tcp` maps the port 2154 on the docker container to 2154 on the host machine. This allows inbound connections on port 2154 to the host machine to be routed to the validator container. In addition to this mapping, it is up to you to ensure that this port is available from the public internet to the host machine. This will require modifying firewall and/or security group settings in your cloud provider of choice.
- `--name validator` names the container, which makes interacting with the docker easier, but feel free to name the container whatever you want.
- `--mount` the parameters above will mount the container's `/var/data/` directory to the systems directory `$HOME/validator_data`. You will want this folder to be persistent across runs of the docker container as it will contain both the blockchain data and the miner key of your Validator.

> - `quay.io/team-helium/validator:latest-val-amd64` Lastly, this points to the docker **_image_** for the Validator that is tagged as the "latest" for the amd64 architecture. For arm architectures, replace with `latest-val-arm64`.

Reviewer 2: ```suggestion
- `quay.io/team-helium/validator:latest-validator-amd64` Lastly, this points to the docker **_image_** for the Validator that is tagged as the "latest" for the amd64 architecture. For arm architectures, replace with `latest-validator-arm64`.
```

Additional flags that may be helpful:

- `-e "NAT_INTERNAL_IP=192.168.0.139"` If your host machine is behind a firewall that is performing network address translation (NAT), adding this flag will assist the Validator to quickly determine the IP address of the host machine. Replace the IP with yours from `ip addr`. This goes hand-in-hand with the following flag:
- `-e "NAT_EXTERNAL_IP=$( curl -s ipv4.icanhazip.com )"` This also assists the Validator in quickly identifying the external IP of the NAT. The `curl` commmand simply queries an external provider and returns the public IP address of the interface. You can remove this and replace with your external IP if it is known and will not change (static IP).
- Similar commands are available with `NAT_INTERNAL_PORT` and `NAT_EXTERNAL_PORT` if any port translations are occuring. If not, these may be omitted.

Once you run the command above, docker will retrieve the latest Validator image, and start the docker process. Barring any errors, your Validator is running.

#### Configure your Docker to update the Validator automatically - RECOMMENDED

> Having all Mainnet Validators running the latest release of Validator software is crucial for the security and stability of the Network. Out-of-date Validators can increase election times or even stall the chain depending on the changes made.

Reviewer 2: ```suggestion
Having all Mainnet Validators running the latest release of Validator software is crucial for the security and stability of the Network. Out-of-date Validators can increase election times or even stall the chain depending on the changes made and possibly incur penalties that would negatively impact earnings.
```

To ensure that each Validator gets updated quickly without your manual involvement, we will use a docker container process called [Watchtower](https://containrrr.dev/watchtower/). This is a docker container that, based on an interval, reviews the running containers to see if there is an updated image available for the container. If so, it stops the running container, and creates a new container with the updated image.

:::info

> Watchtower simply reviews the `docker run` commands that were used to launch the running containers and checks for newer images for these containers.  Therefore it is imperative that your `docker run` command reference the `latest-val-amd64` or `latest-val-arm64` image tag or **it will never update**.

Reviewer 2: ```suggestion
Watchtower simply reviews the `docker run` commands that were used to launch the running containers and checks for newer images for these containers.  Therefore it is imperative that your `docker run` command reference the `latest-validator-amd64` or `latest-validator-arm64` image tag or **it will never update**.
```

Finally, you may be asking if there's a method to know if watchtower was successful in identifying an upgrade and installing it.  Read through the [Watchtower notifications options](https://containrrr.dev/watchtower/notifications/) page how you can extend this guide further and receive notifications through Slack, Teams, Discord, Pushover, email or other methods.

:::info

If you choose to not run Watchtower or something like it for automatic updates of the Docker image, please watch and respond to upgrade requests on the `#validators-announcements` channel on the [Helium Discord](https://discord.gg/helium) server in a timely manner.

:::

#### Upgrade your Docker container manually - Not needed if running watchtower

If you're planning to do manual upgrades, the latest miner tag [can be found here](https://quay.io/repository/team-helium/validator?tag=latest&tab=tags).

To upgrade your docker container to the latest version, do the following. These commands will stop the running container then delete it; pull ann updated Validator docker image from the repository; and finally delete any out of date Validator docker images.

```
docker stop validator && docker rm validator

> docker pull quay.io/team-helium/validator:latest-val-amd64

Reviewer 2: ```suggestion
docker pull quay.io/team-helium/validator:latest-validator-amd64
```

Next, get the `libwxgtk` package:

```
wget http://mirrors.kernel.org/ubuntu/pool/universe/w/wxwidgets3.0/libwxgtk3.0-0v5_3.0.4+dfsg-3_amd64.deb
```

And finish things off by acquiring a some wonderful dependencies:

```
sudo apt install -y libdbus-1-dev autoconf automake libtool flex libgmp-dev cmake libsodium-dev libssl-dev bison libsnappy-dev libclang-dev doxygen make cargo g++ libsctp1 libncurses5 libwxbase3.0-0v5 build-essential cmake libdbus-1-dev mosh vim parallel
```

Then we need to unpack and install all of this stuff. Start with `libwxgtk`:

```
sudo dpkg -i libwxgtk3.0-0v5_3.0.4+dfsg-3_amd64.deb
sudo apt update -y
```

Then onto Erlang. First `unpack` then `install`:

```
sudo dpkg -i esl-erlang_22.3.1-1~ubuntu~bionic_amd64.deb
```

```
sudo apt-get install -f
```

Then navigate to your `miner` directory:

```
cd miner
```

Build the `miner`

```
./rebar3 as validator release
```

Then start the `miner`:

```
_build/validator/rel/miner/bin/miner start
```

#### Load the Genesis Block

One last step. When building from source, you'll need to manually load the Helium blockchain genesis block once your miner is deployed and running. To do this, run the following:

```

> wget https://snapshots.helium.wtf/genesis.testnet

Reviewer 2: ```suggestion
wget https://github.com/helium/blockchain-api/raw/master/priv/prod/genesis
```

+------------------------------------------------+-------------+----------+---------+---+----------+

+----------------------------+
| listen_addrs (prioritized) |
+----------------------------+
|/ip4/173.230.156.39/tcp/2154|
+----------------------------+

+-----------------+--------------------+---------------------------------------+-------------------+
|      local      |       remote       |                  p2p                  |       name        |
+-----------------+--------------------+---------------------------------------+-------------------+
|/ip4/172.17.0.2/t|/ip4/50.16.94.64/tcp|/p2p/1Y9MoitLaEbXPdrZZogURSNQ54eXcp2Ljv|agreeable-tweed-pir|
|/ip4/172.17.0.2/t|/ip4/143.110.235.209|/p2p/1YAFBayNk8bXkkEpqXdxu73kFTjtMqXKTy|prehistoric-hemp-ra|
|/ip4/172.17.0.2/t|/ip4/73.2.34.208/tcp|/p2p/1YEC3g9Hep4hzbDwLFbJ4SqrthTcjGDWdP|wonderful-arctic-ch|
|/ip4/172.17.0.2/t|/ip4/172.90.214.198/|/p2p/1YLJC8DSN6SF17S9nXfYEwTXKtoXpgqBSX|howling-ultraviolet|
|/ip4/172.17.0.2/t|/ip4/13.59.168.136/t|/p2p/1YZhCNPj181YL21aHECXLS1zvDzVY43px9|macho-chocolate-but|
+-----------------+--------------------+---------------------------------------+-------------------+
```

> You're looking for `listen_addrs`. If you don't have at least one, your validator hasn't settled on how to tell other peers to reach it. Often this can take 15-30min, sometimes longer.

Reviewer 2: ```suggestion
You're looking for `listen_addrs`. If you don't have at least one, your Validator hasn't settled on how to tell other peers to reach it. Often this can take 15-30 mins, sometimes longer.
```

### Check the Explorer

> Head over to the [Validator Mainnet Explorer](https://explorer-beta.helium.com/validators) and search for your validator. This may take 10 to 15 minutes from the time that your Validator is both online and staked. You should see your 3-word validator name (`short-umber-bull` in the example above) listed. It may be easier to sort by `#`, descending, as your new Validator will be near the top of the list. Remember you can view the name of your validator through `miner info summary` or just `miner info name`. 

Reviewer 2: ```suggestion
Head over to the [Validator Mainnet Explorer](https://explorer-beta.helium.com/validators) and search for your Validator. This may take 10 to 15 minutes from the time that your Validator is both online and staked before it appears on explorer. You should see your 3-word validator name (`short-umber-bull` in the example above) listed. It may be easier to sort by `#`, descending, as your new Validator will be near the top of the list. Remember you can view the name of your validator through `miner info summary` or just `miner info name`. 
```
